File: mvp/src/main/java/com/android/oleksandrpriadko/mvp/repo_extension/RetrofitRepoExtension.java
package com.android.oleksandrpriadko.mvp.repo_extension;

import android.support.annotation.NonNull;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Converter;
import retrofit2.Retrofit;

public abstract class RetrofitRepoExtension implements RepoExtension {

    private final String mBaseUrl;
    private Retrofit mRetrofit;

    public RetrofitRepoExtension(@NonNull final String baseUrl) {
        this.mBaseUrl = baseUrl;
        this.initRetrofit();
    }

    protected abstract Interceptor[] getInterceptors();

    @NonNull
    protected abstract HttpLoggingInterceptor.Level getLoggingLevel();

    protected abstract Converter.Factory getConverterFactory();

    private void initRetrofit() {
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(getLoggingLevel());
        OkHttpClient.Builder builder = new OkHttpClient.Builder().addInterceptor(interceptor);
        Interceptor[] interceptors = getInterceptors();
        for (int i = 0; i < interceptors.length; ++i) {
            builder.addInterceptor(getInterceptors()[i]);
        }

        OkHttpClient client = builder.build();
        this.mRetrofit = new Retrofit.Builder()
            .baseUrl(getBaseUrl())
            .client(client)
            .addConverterFactory(getConverterFactory())
            .build();
    }

    private String getBaseUrl() {
        return this.mBaseUrl;
    }

    private Retrofit getRetrofit() {
        if (this.mRetrofit == null) {
            this.initRetrofit();
        }
        return this.mRetrofit;
    }

    public <A> A getApi(final Class<A> apiClass) {
        return this.getRetrofit().create(apiClass);
    }

    @Override
    public void cleanUp() {
        mRetrofit = null;
    }

    public interface Listener {

        void onLoadingStarted();

        void onLoadingDone();

        void onLoadingError(@NonNull Throwable throwable);

    }
}

File: mvp/src/main/java/com/android/oleksandrpriadko/mvp/repo_extension/RetrofitRepoExtension.kt
package com.android.oleksandrpriadko.mvp.repo_extension

import android.content.Context
import com.android.oleksandrpriadko.core.CoreServiceManager
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Converter
import retrofit2.Retrofit
import java.util.concurrent.TimeUnit

class RetrofitRepoExtension(
        private val context: Context,
        private val baseUrl: String,
        private val interceptors: List<Interceptor> = listOf(),
        private val loggingLevel: HttpLoggingInterceptor.Level = HttpLoggingInterceptor.Level.NONE,
        private val converterFactory: Converter.Factory) : RepoExtension {

    private var retrofit: Retrofit? = null
        get() {
            if (field == null) {
                initRetrofit()
            }
            return field
        }

    init {
        initRetrofit()
    }

    private fun initRetrofit() {
        val interceptor = HttpLoggingInterceptor()
        interceptor.level = loggingLevel
        val builder = OkHttpClient.Builder()
                .connectTimeout(20, TimeUnit.SECONDS)
                .readTimeout(20, TimeUnit.SECONDS)
                .writeTimeout(20, TimeUnit.SECONDS)
                .retryOnConnectionFailure(false)
                .addInterceptor(interceptor)
        val interceptors = interceptors
        for (i in interceptors.indices) {
            builder.addInterceptor(interceptors[i])
        }

        val client = builder.build()
        retrofit = Retrofit.Builder()
                .baseUrl(baseUrl)
                .client(client)
                .addConverterFactory(converterFactory)
                .build()
    }

    /**
     * return null if there is no internet connection
     */
    fun <T> getApi(apiClass: Class<T>): T? {
        return if (CoreServiceManager.isOnline(context)) {
            retrofit!!.create(apiClass)
        } else {
            null
        }
    }

    override fun cleanUp() {
        retrofit = null
    }

    override fun enableLog(): Boolean {
        return false
    }

    companion object {
        const val RESPONSE_NOT_SUCCESSFUL = "Response is not successful"
        const val RESPONSE_NULL = "Response content is null"
        const val RESPONSE_EMPTY = "Response content is empty"
    }

    interface Listener {

        fun onLoadingStarted()

        fun onLoadingDone()

        fun onLoadingError()

        fun onNoInternet()

    }
}